pkg/services/store/resolver/ds_cache.go

```diff
@@ -0,0 +1,135 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins/manager/registry"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/store"
+	"github.com/grafana/grafana/pkg/tsdb/grafanads"
+)
+
+type dsVal struct {
+	InternalID   int64
+	IsDefault    bool
+	Name         string
+	Type         string
+	UID          string
+	PluginExists bool // type exists
+}
+
+type dsCache struct {
+	ds             datasources.DataSourceService
+	pluginRegistry registry.Service
+	cache          map[int64]map[string]*dsVal
+	timestamp      time.Time // across all orgIDs
+	mu             sync.Mutex
+}
+
+func (c *dsCache) check(ctx context.Context) error {
+	old := c.timestamp
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.timestamp != old {
+		return nil // already updated while we waited!
+	}
+
+	cache := make(map[int64]map[string]*dsVal, 0)
+	defaultDS := make(map[int64]*dsVal, 0)
+
+	q := &datasources.GetAllDataSourcesQuery{}
+	err := c.ds.GetAllDataSources(ctx, q)
+	if err != nil {
+		return err
+	}
+
+	for _, ds := range q.Result {
+		val := &dsVal{
+			InternalID: ds.Id,
+			Name:       ds.Name,
+			UID:        ds.Uid,
+			Type:       ds.Type,
+			IsDefault:  ds.IsDefault,
+		}
+		_, ok := c.pluginRegistry.Plugin(ctx, val.Type)
+		val.PluginExists = ok
+
+		orgCache, ok := cache[ds.OrgId]
+		if !ok {
+			orgCache = make(map[string]*dsVal, 0)
+			cache[ds.OrgId] = orgCache
+		}
+
+		orgCache[val.UID] = val
+
+		// Empty string or
+		if val.IsDefault {
+			defaultDS[ds.OrgId] = val
+		}
+	}
+
+	for orgID, orgDSCache := range cache {
+		// modifies the cache we are iterating over?
```

Comment:
https://go.dev/ref/spec#For_range#:~:text=the%20iteration%20order

> The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next. If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced. If a map entry is created during iteration, that entry may be produced during the iteration or may be skipped. The choice may vary for each entry created and from one iteration to the next. If the map is nil, the number of iterations is 0.

I guess this will work in this case, it just might run more iterations than necessary